File: Chapter 8 - Lecture/p20.m
% program20 - 2nd-order wave equation in 2D via FFT
%             compare program19.m

close all;
scrsz = get(groot,'ScreenSize');
% figure('position', [100   100   4/5*scrsz(3)  4/5*scrsz(4)]);

% Grid and initial data:
N = 48; x = cos(pi*(0:N)/N); y = x';
dt = 6.6/N^2;
[xx,yy] = meshgrid(x,y);
plotgap = round((1/3)/dt); dt = (1/3)/plotgap;
vv = exp(-40*((xx-0.4).^2 + yy.^2));
% The initial velocity is assumed to be almost zero and backward Euler
% formula is used for the first-time approximation. Otherwise, we should
% have special formula to compute u(x,-dt) and u(x,dt) at the first step.
vvold = vv;
% Time-stepping by leap frog formula:
[ay, ax] = meshgrid([0.56 0.06], [0.1 0.55]); clf
for n = 0:3*plotgap
    t = n*dt;
    if rem(n+0.5,plotgap) < 1 % plots at multiples of t = 1/3
        i = n/plotgap+1;
        % subplot('position', [ax(i) ay(i) 0.36 0.36]);
        figure(i)
        [xxx,yyy] = meshgrid(-1:1/16:1,-1:1/16:1);
        vvv = interp2(xx,yy,vv,xxx,yyy,'spline');
        mesh(xxx,yyy,vvv), axis([-1 1 -1 1 -0.15 1])
        colormap([0 0 0]), title(['t = ' num2str(t)], 'FontSize', 20), drawnow
    end
    uxx = zeros(N+1,N+1); uyy = zeros(N+1,N+1);
    ii = 2:N;
    for i = 2:N   % 2nd derivs wrt x in each row
        v = vv(i,:); V = [v fliplr(v(ii))];
        U = real(fft(V));
        W1 = real(ifft(1i*[0:N-1 0 1-N:-1].*U));	% diff wrt theta
        W2 = real(ifft(-[0:N 1-N:-1].^2.*U));       % diff^2 wrt theta
        uxx(i,ii) = W2(ii)./(1-x(ii).^2) - x(ii).*...
            W1(ii)./(1-x(ii).^2).^(3/2);
    end
    for j = 2:N   % 2nd derivs wrt y in each row
        v = vv(:,j); V = [v; flipud(v(ii))];
        U = real(fft(V));
        W1 = real(ifft(1i*[0:N-1 0 -N+1:-1]'.*U));  % diff wrt theta
        W2 = real(ifft(-[0:N 1-N:-1]'.^2.*U));      % diff^2 wrt theta
        uyy(ii,j) = W2(ii)./(1-y(ii).^2) - y(ii).*...
            W1(ii)./(1-y(ii).^2).^(3/2);
    end
    vvnew = 2*vv - vvold + dt^2*(uxx + uyy);
    vvold = vv; vv = vvnew;
end

File: p20.py
"""
Solve 2nd-order wave equation in 2D via FFT

@author: Khiem Nguyen
"""
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import axes3d
from scipy.interpolate import interp2d
from numpy.fft import fft, ifft


N = 24
x = np.cos(np.pi * np.linspace(0, 1, N+1))
y = x
dt = 6.0/ N**2
xx, yy = np.meshgrid(x, x)

num_snapshots = 4;

plotgap = int((1/(num_snapshots - 1) / dt))
dt = (1/(num_snapshots - 1))/plotgap
vv = np.exp(-40*( (xx - 0.4)**2 + yy**2))
vvold = vv

plt.close('all')

plt.figure(figsize= (10, 10), dpi=80)
# Time-stepping by leap-frog formula
for n in range((num_snapshots-1) * plotgap):
    # t = n* dt
    if n % plotgap == 0 or n == (num_snapshots-1)*plotgap - 1:
        ax = plt.subplot(2,2, (n+1)//plotgap + 1, projection = '3d')
        v_interp = interp2d(xx, yy, vv, kind = 'quintic')
        xxx = np.linspace(-1, 1, 41)
        vvv = v_interp(xxx, xxx)
        xxx, yyy = np.meshgrid(xxx, xxx)
        ax.plot_wireframe(xxx, yyy, vvv, color=[0, 0, 0])
        ax.set_zlim(-0.5, 1)
        
        
    # Compute solution by leap-forg stepping
    uxx = np.zeros((N+1, N+1))
    uyy = np.zeros((N+1, N+1))
    idx = np.arange(1, N)
    k = np.fft.fftfreq(2*N, 1/(2*N)); k[N] = 0
    k2 = np.fft.fftfreq(2*N, 1/(2*N)) ** 2
    jk = 1j*k
    for i in range(1, N):
        v = vv[i, :]
        V = np.append(v, np.flip(v[idx]))
        U = np.real(fft(V))
        
        W1 = np.real(ifft(jk * U))      # 1st-derivative
        W2 = np.real(ifft(-k2 * U))     # 2nd-derivative
        uxx[i, idx] = W2[idx] / (1 - x[idx]**2) - x[idx] * W1[idx]/((1 -x[idx]**2)**1.5)
    
    for j in range(1, N):
        v = vv[:, j]
        V = np.append(v, np.flip(v[idx]))
        U = np.real(fft(V))
        W1 = np.real(ifft(jk * U))
        W2 = np.real(ifft(-k2 * U))
        uyy[idx, j] = W2[idx] / (1 - y[idx]**2) - y[idx] * W1[idx]/((1 - y[idx]**2)**1.5)
    
    vvnew = 2*vv - vvold + dt**2 * (uxx + uyy)
    vvold = vv
    vv = vvnew
    
v_val = vv[N//2, N//2]
plt.show()
